core/src/main/java/com/linecorp/armeria/client/RefreshingAddressResolverGroup.java

```diff
@@ -137,16 +140,19 @@
     @Override
     public void close() {
         super.close();
-        while (!cache.isEmpty()) {
-            for (final Iterator<Entry<String, CompletableFuture<CacheEntry>>> i = cache.entrySet().iterator();
-                 i.hasNext();) {
-                final Entry<String, CompletableFuture<CacheEntry>> entry = i.next();
-                i.remove();
-                entry.getValue().handle((cacheEntry, cause) -> {
+        cache.invalidateAll();
+    }
+
+    private static Cache<String, CompletableFuture<CacheEntry>> buildCache(String cacheSpec) {
+        final Caffeine<Object, Object> b = Caffeine.from(cacheSpec);
+        b.removalListener((RemovalListener<String, CompletableFuture<CacheEntry>>) (key, value, cause) -> {
+            if (value != null) {
+                value.handle((cacheEntry, throwable) -> {
                     cacheEntry.clear();
```

Comment:
I guess this `RemovalListener` could be called by any thread if the `value` is completed already and `close()` is called by a user thread.
Don't we need to handle the concurrency issue in the `CacheEntry.clear()`?
For example:
```java
void clear() {
	if (executor().inEventLoop()) {
       clear0()
    } else {
       executor().execute(this::clear0)
    }
}
```